File: app/src/main/java/com/example/movies/MovieAdapter.java
package com.example.movies;

import android.graphics.drawable.Drawable;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import java.util.ArrayList;
import java.util.List;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {

    private List<Movie> movieList = new ArrayList<>();
    private onReachEndListener onReachEndListener;
    private final static String TAG = "MovieAdapter";

    public void setOnReachEndListener(MovieAdapter.onReachEndListener onReachEndListener) {
        this.onReachEndListener = onReachEndListener;
    }

    public void setMovies(List<Movie> movies) {
        this.movieList = movies;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.movie_item, parent, false);
        return new MovieViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        Log.d(TAG, "onBindViewHolder: " + position);
        Movie movie = movieList.get(position);
        Glide.with(holder.itemView)
                .load(movie.getPoster().getUrl())
                .into(holder.imageViewPoster);

        double rating = movie.getRating().getKp();
        int backgroundId;
        if(rating > 7) {
            backgroundId = R.drawable.circle_green;
        } else if(rating > 5) {
            backgroundId = R.drawable.circle_orange;
        } else {
            backgroundId = R.drawable.circle_red;
        }
        Drawable background = ContextCompat.getDrawable(holder.itemView.getContext(), backgroundId);
        holder.textViewRating.setBackground(background);
        holder.textViewRating.setText(String.valueOf(rating).substring(0, 3));

        if(movieList.size() - 10 <= position && onReachEndListener != null) {
            onReachEndListener.onReachEnd();
        }
    }

    @Override
    public int getItemCount() {
        return movieList.size();
    }

    interface onReachEndListener {
        void onReachEnd();
    }

    static class MovieViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageViewPoster;
        private final TextView textViewRating;

        public MovieViewHolder(@NonNull View itemView) {
            super(itemView);
            imageViewPoster = itemView.findViewById(R.id.imageViewPoster);
            textViewRating = itemView.findViewById(R.id.textViewRating);
        }
    }
}

File: app/src/main/java/com/example/movies/adapters/MovieAdapter.kt
package com.example.movies.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.bumptech.glide.Glide
import com.example.movies.R
import com.example.movies.databinding.MovieItemBinding
import com.example.movies.pojo.Movie

class MovieAdapter : RecyclerView.Adapter<MovieAdapter.MovieViewHolder>() {

    private var movieList: List<Movie> = listOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var onReachEndListener: OnReachEndListener? = null
    var onOpenScreenDetailListener: OnOpenScreenDetailListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val view = MovieItemBinding.inflate(
            LayoutInflater
                .from(parent.context), parent, false
        )
        return MovieViewHolder(view)
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        with(holder) {
            with(binding) {
                val movie = movieList[position]
                Glide.with(itemView)
                    .load(movie.poster.url)
                    .into(imageViewPoster)

                val rating = movie.rating.kp
                val backgroundId = when {
                    rating > 7 -> R.drawable.circle_green
                    rating > 5 -> R.drawable.circle_orange
                    else -> R.drawable.circle_red
                }
                val background = ContextCompat.getDrawable(itemView.context, backgroundId)
                textViewRating.background = background
                textViewRating.text = rating.toString().substring(0, 3)

                if (movieList.size - 10 <= position) onReachEndListener?.onReachEnd()
                imageViewPoster.setOnClickListener {
                    onOpenScreenDetailListener?.onOpenScreenDetail(movie)
                }
            }

        }
    }

    override fun getItemCount() = movieList.size

    interface OnReachEndListener {
        fun onReachEnd()
    }

    interface OnOpenScreenDetailListener {
        fun onOpenScreenDetail(movie: Movie)
    }

    inner class MovieViewHolder(val binding: MovieItemBinding) : ViewHolder(binding.root)
}